maven-core/src/main/java/org/apache/maven/artifact/handler/manager/DefaultArtifactHandlerManager.java

```diff
@@ -22,51 +22,55 @@
 import javax.inject.Named;
 import javax.inject.Singleton;
 
+import java.util.Collections;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 
+import org.apache.maven.api.Type;
+import org.apache.maven.api.services.TypeRegistry;
 import org.apache.maven.artifact.handler.ArtifactHandler;
 import org.apache.maven.artifact.handler.DefaultArtifactHandler;
 
+import static java.util.Objects.requireNonNull;
+
 /**
  */
 @Named
 @Singleton
 public class DefaultArtifactHandlerManager implements ArtifactHandlerManager {
+    private final TypeRegistry typeRegistry;
 
-    private final Map<String, ArtifactHandler> artifactHandlers;
-
-    private final Map<String, ArtifactHandler> allHandlers = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, ArtifactHandler> handlers;
 
     @Inject
-    public DefaultArtifactHandlerManager(Map<String, ArtifactHandler> artifactHandlers) {
-        this.artifactHandlers = artifactHandlers;
+    public DefaultArtifactHandlerManager(TypeRegistry typeRegistry) {
+        this.typeRegistry = requireNonNull(typeRegistry, "null typeRegistry");
+        this.handlers = new ConcurrentHashMap<>();
     }
 
-    public ArtifactHandler getArtifactHandler(String type) {
-        ArtifactHandler handler = allHandlers.get(type);
-
-        if (handler == null) {
-            handler = artifactHandlers.get(type);
-
-            if (handler == null) {
-                handler = new DefaultArtifactHandler(type);
-            } else {
-                allHandlers.put(type, handler);
-            }
-        }
-
-        return handler;
+    public ArtifactHandler getArtifactHandler(String id) {
+        return handlers.computeIfAbsent(id, k -> {
+            Type type = typeRegistry.getType(id);
+            return new DefaultArtifactHandler(
+                    id,
+                    type.getExtension(),
+                    type.getClassifier(),
+                    null,
+                    null,
+                    type.isIncludesDependencies(),
+                    type.getLanguage(),
+                    type.isAddedToClassPath()); // TODO: watch out for module path
+        });
     }
 
     public void addHandlers(Map<String, ArtifactHandler> handlers) {
         // legacy support for maven-gpg-plugin:1.0
-        allHandlers.putAll(handlers);
+        throw new IllegalArgumentException("you cannot do this anymore");
```

Comment:
`throw new UnsupportedOperationException("Adding handlers is not supported anymore")` ?